lib/hashset.rs

```diff
@@ -0,0 +1,387 @@
+        HashSet::from(self.set.update(x))
+    }
+    pub fn union(self, other: Self) -> Self {
+        // `im::HashSet` implements union by adding elements from `other` to
+        // `self`.  This is inefficient when `other` is much
+        // larger than `self`.  The following optimization will become
+        // unnecessary once this PR or something similar has landed:
+        // https://github.com/bodil/im-rs/pull/163.
+        if self.len() >= other.len() {
+            HashSet::from(self.set.union(other.set))
+        } else {
+            HashSet::from(other.set.union(self.set))
+        }
+    }
+    pub fn unions<I>(i: I) -> Self
+    where
+        I: IntoIterator<Item = Self>,
+    {
+        let mut sets: Vec<_> = i.into_iter().collect();
+
+        // Start with the largest set.
+        if let Some((largest_idx, _)) = sets.iter().enumerate().max_by_key(|(_, s)| s.len()) {
+            let mut largest_set = Self::default();
+            std::mem::swap(&mut largest_set, &mut sets[largest_idx]);
+
+            sets.into_iter().fold(largest_set, Self::union)
+        } else {
+            Self::default()
+        }
+    }
+    pub fn intersection(self, other: Self) -> Self {
+        HashSet::from(self.set.intersection(other.set))
+    }
+    pub fn difference(self, other: Self) -> Self {
+        HashSet::from(self.set.difference(other.set))
+    }
+}
+
+// I don't want to implement `Deref` for `HashSet` to prevent
+// accidentally calling `im::HashSet` methods.  For example,
+// `im::HashSet::iter` returns the non-deterministic iterator (which can
+// mess up DDlog in ways one can easily miss unless they test on big
+// enough sets).
+
+// impl<T> Deref for HashSet<T> {
+//     type Target = IMHashSet<T>;
+//
+//     fn deref(//self) -> //Self::Target {
+//         //self.set
+//     }
+// }
+//
+// impl<T> DerefMut for HashSet<T> {
+//     fn deref_mut(//mut self) -> //mut Self::Target {
+//         //mut self.set
+//     }
+// }
+
+impl<T: FromRecord + Hash + Eq + Clone + Ord> FromRecord for HashSet<T> {
+    fn from_record(val: &Record) -> StdResult<Self, String> {
+        match val {
+            Record::Array(_, args) => StdResult::from_iter(args.iter().map(T::from_record)),
+            v => T::from_record(v).map(Self::unit),
+        }
+    }
+}
+
+impl<T: IntoRecord + Hash + Eq + Clone + Ord> IntoRecord for HashSet<T> {
+    fn into_record(self) -> Record {
+        Record::Array(
+            CollectionKind::Set,
+            self.into_iter().map(|x| x.into_record()).collect(),
+        )
+    }
+}
+
+// Set update semantics: update contains values that are in one of the sets
+// but not the other.
+impl<T: FromRecord + Hash + Eq + Clone + Ord> Mutator<HashSet<T>> for Record {
+    fn mutate(&self, set: &mut HashSet<T>) -> StdResult<(), String> {
+        let upd = <HashSet<T>>::from_record(self)?;
+        for v in upd.into_iter() {
+            if set.remove(&v).is_none() {
+                set.insert(v);
```

Comment:
I don't quite understand the rationale here, why is the element removed and inserted if there wasn't previously a value?

Reply:
It's basically what the above comment says.  `upd` is a set that describes a set of updates to apply to `set`.  For each value in `upd`, if the value is in `set`, then it must be deleted, and if not, then it must be inserted.  Actually, now I realize that this is just the symmetric difference of `upd` and `set`, and we have a function for it, so that's what I should be using instead.